render/gles2/renderer.c

```diff
@@ -31,13 +31,142 @@ static struct wlr_gles2_renderer *gles2_get_renderer_in_context(
 	return renderer;
 }
 
+static void destroy_buffer(struct wlr_gles2_buffer *buffer) {
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	wlr_egl_make_current(buffer->renderer->egl, EGL_NO_SURFACE, NULL);
+
+	PUSH_GLES2_DEBUG;
+
+	glDeleteFramebuffers(1, &buffer->gl_fbo);
+	glDeleteTextures(1, &buffer->gl_texture);
+
+	POP_GLES2_DEBUG;
+
+	wlr_egl_destroy_image(buffer->renderer->egl, &buffer->egl_image);
+	free(buffer);
+}
+
+static struct wlr_gles2_buffer *get_buffer(struct wlr_gles2_renderer *renderer,
+		struct wlr_buffer *wlr_buffer) {
+	struct wlr_gles2_buffer *buffer;
+	wl_list_for_each(buffer, &renderer->buffers, link) {
+		if (buffer->buffer == wlr_buffer) {
+			return buffer;
+		}
+	}
+	return NULL;
+}
+
+static void handle_buffer_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_gles2_buffer *buffer =
+		wl_container_of(listener, buffer, buffer_destroy);
+	destroy_buffer(buffer);
+}
+
+static struct wlr_gles2_buffer *create_buffer(struct wlr_gles2_renderer *renderer,
+		struct wlr_buffer *wlr_buffer) {
+	struct wlr_gles2_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	if (!wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
+		goto error_buffer;
+	}
+
+	bool external_only;
+	buffer->egl_image = wlr_egl_create_image_from_dmabuf(renderer->egl,
+		&dmabuf, &external_only);
+	if (buffer->egl_image == EGL_NO_IMAGE_KHR) {
+		goto error_buffer;
+	}
+
+	PUSH_GLES2_DEBUG;
+
+	glGenTextures(1, &buffer->gl_texture);
+	glBindTexture(GL_TEXTURE_2D, buffer->gl_texture);
+	gles2_procs.glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, buffer->egl_image);
```

Comment:
Considered using `glEGLImageTargetRenderbufferStorageOES` here instead of `glEGLImageTargetTexture2DOES`?

Reply:
Right, it would probably be better since we don't use the texture for sampling.

Reply:
Well, I suppose we could sample the buffer in rare cases such as `wlr_renderer_blit_dmabuf()`. I wonder if there's any performance difference inherent in using renderbuffers vs. texture.

Reply:
Switched to RBOs